README.md
# Server Sent Events
The Server-Sent Events specification describes a built-in class EventSource, that keeps connection with the server and allows to receive events from it.

Similar to WebSocket, the connection is persistent.

But there are several important differences:

| WebSocket                            | EventSource                        |
| ------------------------------------ | ---------------------------------- |
| Bi-directional: both client and server can exchange messages  | One-directional: only server sends data  |
| Binary and text data                                          | Only text  |
| WebSocket protocol                                            | Regular HTTP |

EventSource is a less-powerful way of communicating with the server than WebSocket.

Why should one ever use it?

The main reason: it’s simpler. In many applications, the power of WebSocket is a little bit too much.

We need to receive a stream of data from server: maybe chat messages or market prices, or whatever. That’s what EventSource is good at. Also it supports auto-reconnect, something we need to implement manually with WebSocket. Besides, it’s a plain old HTTP, not a new protocol.

### How to start

- Step-1
```
npm install
```

- Step-2
```
npm start
```

- Step-3: Open below url in the browser
```
http://localhost:5000
```



CHANGELOG.md
# server-sent-events => @v1.0.0
- Added Nodejs Http with Rxjs Observable sample
- Added Hapijs framework sample
- Adde Basic html template


src/hapi-app.js
const { PassThrough } = require('stream');
const Hapi = require('hapi');
const server = new Hapi.Server();
server.connection({
    port: '4000'
});

// server.load = {
//     sampleInterval: 1000
// };

server.route([{
    method: 'GET',
    path: '/',
    config: {
        handler: function (request, reply) {
            return reply(`
            <!DOCTYPE html>
            <html lang="en" dir="ltr">
            <head>
                <meta charset="utf-8">
                <title>SSE</title>
            </head>
            <body>
                <pre id="log"></pre>
            </body>
            <script>
                var eventSource = new EventSource('/events');
                eventSource.onmessage = function(event) {
                document.getElementById('log').innerHTML += event.type + " - " + event.data +'<br>';
                };
                eventSource.addEventListener("update", function(event) {
                    document.getElementById('log').innerHTML += event.type + " - " + event.data +'<br>';
                });
                // eventSource.addEventListener("message", function(e) {
                //     console.log("Message", e.data)
                //   })
                eventSource.onerror = (event) => { 
                    document.getElementById('log').innerHTML += 'Server closes the connection<br>';
                    eventSource.close();
                }
                eventSource.addEventListener("close", function(e) {
                    document.getElementById('log').innerHTML += event.type + " - " + event.data +'<br>';
                    eventSource.close() 
                })
            </script>
            </html>
            `).type('text/html');
        }
    }
},
{
    method: 'GET',
    path: '/events',
    config: {
        // cors: { 
        //     origin: ['*'],
        //     additionalHeaders: ['cache-control', 'x-requested-with']
        // },
        handler: function (request, reply) {
            const channel = new PassThrough();
            const refreshRate = 500; // in milliseconds;
            let eventCount = 0, eventId = 0;
            /* Send Events Format
                @eventType - default type 'message'
                @id - string
                @data - string | object
                @retry - Number in milliseconds
            */
            
            let eventScheduler = setInterval(() => {
                eventCount++;
                eventId = Date.now();
       
                // Message - Event
                if(eventCount === 1) {
                    channel.write(`event:message\n`);
                    channel.write(`id:${eventId}\n`);
                    channel.write(`data: Event, Welcome to learn about server sent events. <span id="liveUpdate1" style="background:yellow;color:black;padding:5px;"></span> \n`);
                    channel.write(`retry:10000\n\n`);
                }

                // Update - Event
                channel.write(`event:update\n`);
                channel.write(`id:${eventId}\n`);
                channel.write(`data:{ "info": "Update Event triggered", "eventCount": ${eventCount} }\n`);
                channel.write(`retry:10000\n\n`);

            }, refreshRate)

            request.raw.req.on("close", function() {
                console.info("[Hapijs App] Connection Closed by Browser");
                channel.end();
                clearInterval(eventScheduler);
            });

            setTimeout(() =>{
                channel.write(`event:close\n`);
                channel.write(`id:${eventCount}\n`);
                channel.write(`data: Event triggered from server to close the connection at client side.\n`);
                channel.write(`retry:10000\n\n`);
                channel.end();
                clearInterval(eventScheduler);
            }, 10000);
            
            /* Native nodejs response without framework */
                //request.raw.res.writeHead(200, { 'content-type': 'text/event-stream; charset=utf-8', 'connection': 'keep-alive', 'cache-control': 'no-cache' });
                //channel.pipe(request.raw.res);

            /* Reply must be set header
                Content-Encoding: identity
            */
            return reply(channel)
                .type('text/event-stream')
                .header('Access-Control-Allow-Origin', '*')
                //.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept')
                .header('Cache-Control','no-cache')
                .header('Connection','keep-alive')
                .header('Content-Encoding', 'identity');
        }
    }
}]);

server.start(function (err) {  
    console.log(`[Hapijs App] Listening on http://localhost:${server.info.port}`);
    // setInterval(() => { 
    //     server.getConnections((error, count) => {
    //         console.log("Open Connection Count", count);
    //     });
    // }, 1000);
});


src/http-app.js
const http = require('http');
const port = process.env.PORT || 3000;
const { PassThrough } = require('stream');
const server = http.createServer((req, res) => {
    if(req.url === '/events') {
        return dispatchEventServerSentEvent(req, res);
    }
    return dispatchHomePage(req, res);
});

server.listen(port);

server.on('error', (err) => {
    console.log(err);
    process.exit(1);
});
  
server.on('listening', () => {
    console.log(`[Http App] Listening on http://localhost:${port}`);
    setInterval(() => { 
        server.getConnections((error, count) => {
            console.log("[Http App] Open Connection Count", count);
        });
    }, 1000);
});


function dispatchEventServerSentEvent(req, res) {
    const channel = new PassThrough();
    const refreshRate = 500; // in milliseconds
    let eventCount = 0, eventId = 0, eventScheduler;
    res.writeHead(200, { 
        'Content-Type': 'text/event-stream',
        'Access-Control-Allow-Origin': '*',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive',
        'Content-Encoding': 'identity'
    });
    
    /* Send Events Format
        @eventType - default type 'message'
        @id - string
        @data - string | object
        @retry - Number in milliseconds
    */
    
    eventScheduler = setInterval(() => {
        eventCount++;
        eventId = Date.now();
       
        // Message - Event
        if(eventCount === 1) {
            channel.write(`event:message\n`);
            channel.write(`id:${eventId}\n`);
            channel.write(`data: Event, Welcome to learn about server sent events. <span id="liveUpdate" style="background:gray;color:white;padding:5px;"></span> \n`);
            channel.write(`retry:10000\n\n`);
        }
        
        // Update - Event
        channel.write(`event:update\n`);
        channel.write(`id:${eventId}\n`);
        channel.write(`data:{ "info": "Update Event triggered", "eventCount": ${eventCount} }\n`);
        channel.write(`retry:10000\n\n`);

    }, refreshRate);

    req.on('close', () => {
        console.log("[Http App] Connection Closed by Browser");
        channel.end();
        clearInterval(eventScheduler);
    })

    setTimeout(() => {
    // Close - Event
        channel.write(`event:close\n`);
        channel.write(`id:${eventId}\n`);
        channel.write(`data: Event triggered from server to close the connection at client side.\n`);
        channel.write(`\n`); // end of stream
        channel.end();
        clearInterval(eventScheduler);
    }, 10000);

    return channel.pipe(res);

}

function dispatchHomePage(req, res) {
    res.writeHead(200, { 'Content-Type': 'text/html' });
    res.end(`
    <!DOCTYPE html>
    <html lang="en" dir="ltr">
      <head>
        <meta charset="utf-8">
        <title>SSE</title>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/rxjs/6.6.7/rxjs.umd.min.js" integrity="sha512-0/2ebe9lI6BcinFBXFjbBkquDfccT2wP+E48pecciFuGMXPRNdInXZawHiM2NUUVJ4/aKAzyebbvh+CkvRhwTA==" crossorigin="anonymous"></script>
        </head>
      <body>
      <div id="log"></div>

      </body>
      <script>
        const { Observable, Subject, fromEvent, operators } = rxjs;
        
        const eventSource = new EventSource('/events');

        const element = document.querySelector('#log');
        fromEvent(eventSource, 'message').subscribe(event => {
            element.innerHTML += '<u>' + event.type.toUpperCase() + '</u>' + ': '+ event.data +'<br><br>';
        });
        console.log(operators);
        fromEvent(eventSource, 'update')
        .pipe(
            operators.take(20),
            operators.tap(event => console.log(event)),
            //operators.debounceTime(1000)
            )
        .subscribe(event => {
            document.querySelector('#liveUpdate').innerHTML = event.lastEventId;
            element.innerHTML += '<u>' + event.type.toUpperCase() + '</u>' + ': Event' + '<br>Last Event ID: ' + event.lastEventId + "<br>DATA: " + event.data +'<br><br>';
        });

        fromEvent(eventSource, 'close').subscribe(event => {
            element.innerHTML += '<u>' + event.type.toUpperCase() + '</u>' + ': '+ event.data +'<br><br>';
            eventSource.close();
        });

        fromEvent(eventSource, 'error').subscribe(event => {
            element.innerHTML += '<u>' + 'Error:' + '</u>' + 'Event, Server closes the connection.';
            eventSource.close();
        });
      </script>
    </html>
    `);
}


index.js
const http = require('http');
const port = process.env.PORT || 5000;
require('./src/http-app');
require('./src/hapi-app');
const server = http.createServer(dispatchHomePage);
server.listen(port);
server.on('listening', () => {
    console.log(`[UI App] Listening on http://localhost:${port}`);
});
function dispatchHomePage(req, res) {
    res.writeHead(200, { 
        'Content-Type': 'text/html',
        'Access-Control-Allow-Origin': '*'
    });
    res.end(`
    <!DOCTYPE html>
    <html lang="en" dir="ltr">
      <head>
        <meta charset="utf-8">
        <title>NJS: Sever Sent Events(SSE) - Demo</title>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/rxjs/6.6.7/rxjs.umd.min.js" integrity="sha512-0/2ebe9lI6BcinFBXFjbBkquDfccT2wP+E48pecciFuGMXPRNdInXZawHiM2NUUVJ4/aKAzyebbvh+CkvRhwTA==" crossorigin="anonymous"></script>
        <style>    
            .content {
                padding:5px;
            }
            thead,
            tfoot {
                background-color: #3f87a6;
                color: #fff;
                text-align: center;
            }

            tbody {
                background-color: #e4f0f5;
            }

            caption {
                padding: 10px;
                caption-side: bottom;
            }

            table {
                border-collapse: collapse;
                border: 2px solid rgb(200, 200, 200);
                letter-spacing: 1px;
                font-family: sans-serif;
                font-size: .8rem;
                margin:auto;
            }
        </style>
      </head>
      <body>
        <div class="content">
            <p style="text-align:center;"><u><b>Nodejs - Sever Sent Events(SSE) - Demo</b></u></p>
        <table border='1'>    
            <thead>
                <tr>
                    <td>
                        HTTP with RxJs Observable
                    </td>
                    <td>
                        HAPI JS
                    </td>
                </tr>
             </thead> 
             <tbody>
                <tr>
                    <td style="vertical-align:top;padding:5px;">
                        <div id="log"></div>
                    </td>
                    <td style="vertical-align:top;padding:5px;">
                        <div id="log1"></div>
                    </td>
                </tr>
            </tbody>
        </table>
        </div>
      </body>
      <script>
        /*** Http Event Source Listener ***/
        const { Observable, Subject, fromEvent, operators } = rxjs;
        const HttpAppEventSource = new EventSource('http://localhost:3000/events');

        const element = document.querySelector('#log');
        fromEvent(HttpAppEventSource, 'message').subscribe(event => {
            element.innerHTML += '<u>' + event.type.toUpperCase() + '</u>' + ': '+ event.data +'<br><br>';
        });
        fromEvent(HttpAppEventSource, 'update')
        .pipe(
            operators.take(10),
            operators.tap(event => console.log(event)),
            //operators.debounceTime(1000)
            )
        .subscribe(HttpAppEventSource => {
            document.querySelector('#liveUpdate').innerHTML = event.lastEventId;
            element.innerHTML += '<u>' + event.type.toUpperCase() + '</u>' + ': Event' + '<br>Last Event ID: ' + event.lastEventId + "<br>DATA: " + event.data +'<br><br>';
        });

        fromEvent(HttpAppEventSource, 'close').subscribe(event => {
            element.innerHTML += '<u>' + event.type.toUpperCase() + '</u>' + ': '+ event.data +'<br><br>';
            HttpAppEventSource.close();
        });

        fromEvent(HttpAppEventSource, 'error').subscribe(event => {
            element.innerHTML += '<u>' + 'Error:' + '</u>' + 'Event, Server closes the connection.';
            HttpAppEventSource.close();
        });
        
        /*** Hapijs Event Source Listener ***/
        var HapiJsAppEventSource = new EventSource('http://localhost:4000/events');
        const element1 = document.querySelector('#log1');
        HapiJsAppEventSource.onmessage = function(event) {
            element1.innerHTML += '<u>' + event.type.toUpperCase() + '</u>' + ': '+ event.data +'<br><br>';
        };
        HapiJsAppEventSource.addEventListener("update", function(event) {
            document.querySelector('#liveUpdate1').innerHTML = event.lastEventId;
            element1.innerHTML += '<u>' + event.type.toUpperCase() + '</u>' + ': Event' + '<br>Last Event ID: ' + event.lastEventId + "<br>DATA: " + event.data +'<br><br>';
        });
        HapiJsAppEventSource.addEventListener("close", function(e) {
            element1.innerHTML += '<u>' + event.type.toUpperCase() + '</u>' + ': '+ event.data +'<br><br>';
            HapiJsAppEventSource.close();
        });

        HapiJsAppEventSource.onerror = (event) => { 
            element1.innerHTML += '<u>' + 'Error:' + '</u>' + 'Event, Server closes the connection.';
            HapiJsAppEventSource.close();
        }
      </script>
    </html>
    `);
}
